### PR TITLE
Implement category model

### PR DESCRIPTION
**Pull request by Parashchuk**

### CODE REVIEW

**Review by pavlov-n — APPROVED** [review]
Looks good

File: src/main/java/bookstore/entity/category/Category.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false, unique = true)
+    private String name;
+
+    @Column(nullable = false)
+    private boolean isDeleted;
+    private String description;
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/entity/category/Category.java]:
```suggestion
     private String description;
     
    @Column(nullable = false)
    private boolean isDeleted;
```

File: src/main/java/bookstore/mapper/BookMapper.java [hunk 2]
```diff
@@ -1,22 +1,23 @@
...
         injectionStrategy = InjectionStrategy.CONSTRUCTOR,
         nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
         unmappedTargetPolicy = ReportingPolicy.IGNORE,
         implementationPackage = "<PACKAGE_NAME>.impl"
 )
 public interface BookMapper {
-    BookResponseDto toResponseDto(Book book);
+    BookResponseDto toDto(Book book);
 
     Book toModel(BookRequestDto bookRequestDto);
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/mapper/BookMapper.java]: Modify the BookMapper class to have such methods:
BookDto toDto(Book book);
Book toEntity(CreateBookRequestDto bookDto);
BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book); (HINT: BookDtoWithoutCategoryIds class could be used as a response class for @GetMapping("/{id}/books") endpoint)
@AfterMapping default void setCategoryIds(@MappingTarget BookDto bookDto, Book book)

File: src/main/java/bookstore/service/BookService.java [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
 
-import bookstore.dto.BookRequestDto;
-import bookstore.dto.BookResponseDto;
+import bookstore.dto.book.BookRequestDto;
+import bookstore.dto.book.BookResponseDto;
 import java.util.List;
 import org.springframework.data.domain.Pageable;
 
 public interface BookService {
     BookResponseDto save(BookRequestDto book);
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/BookService.java:9]: use BookDtoWithoutCategoryIds (check other methods)

File: src/main/java/bookstore/service/impl/BookServiceImpl.java [hunk 4]
```diff
@@ -20,13 +20,20 @@ public class BookServiceImpl implements BookService {
 
     @Override
     public BookResponseDto save(BookRequestDto bookRequestDto) {
-        return bookMapper.toResponseDto(bookRepository.save(bookMapper.toModel(bookRequestDto)));
+        return bookMapper.toDto(bookRepository.save(bookMapper.toModel(bookRequestDto)));
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/impl/BookServiceImpl.java]: Can you save the book with categories? Let’s  fix it ;)

File: src/main/java/bookstore/service/impl/CategoryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Override
+    public CategoryDto updateById(Long id, CategoryDto categoryDto) {
+        Category category = categoryRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Category wasn't updated, because category with id: " + id
+                                + " doesn't exist"
+                )
+        );
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/impl/CategoryServiceImpl.java:51]: you can use existById

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve  your solution ;)

File: src/main/java/bookstore/service/BookService.java [hunk 6]
```diff
@@ -1,13 +1,15 @@
...
 
+    List<BookResponseDto> findAllByCategoryId(Long id, Pageable pageable);
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/BookService.java]: Can you use BookDtoWithoutCategoryIds? (check other methods)

File: src/main/java/bookstore/controller/BookController.java [hunk 7]
```diff
@@ -48,21 +68,13 @@ public BookResponseDto getBookById(@PathVariable @Positive Long id) {
...
-        return bookService.save(book);
-    }
-
     @PutMapping("/{id}")
     @ResponseStatus(HttpStatus.OK)
     @PreAuthorize("hasRole('ADMIN')")
     @Operation(description = "Update a book by its ID")
     public BookResponseDto updateBookById(
             @PathVariable @Positive Long id,
-            @Valid @RequestBody BookRequestDto book
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/controller/BookController.java:65]: UpdateBookRequestDto

File: src/main/java/bookstore/controller/BookController.java [hunk 8]
```diff
@@ -32,6 +33,14 @@
 public class BookController {
     private final BookService bookService;
 
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(description = "Create a new book")
+    public BookResponseDto createBook(@RequestBody @Valid BookCreateDto book) {
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/controller/BookController.java]: CreateBookCreateDto

File: src/main/java/bookstore/mapper/BookMapper.java [hunk 9]
```diff
@@ -1,25 +1,30 @@
...
 
-    Book toModel(BookRequestDto bookRequestDto);
+    @Mapping(target = "categories", ignore = true)
+    Book toModel(BookCreateDto bookCreateDto);
 
-    void updateBook(BookRequestDto dto, @MappingTarget Book book);
+    @Mapping(target = "categories", ignore = true)
+    void updateBook(BookCreateDto dto, @MappingTarget Book book);
 }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/mapper/BookMapper.java]: leave only one empty line

File: src/main/java/bookstore/repository/BookRepository.java [hunk 10]
```diff
@@ -10,4 +11,7 @@
 public interface BookRepository extends JpaRepository<Book, Long> {
     @EntityGraph(attributePaths = "categories")
     List<Book> findAllByCategoriesId(Long id);
+
+    @EntityGraph(attributePaths = "categories")
+    List<Book> findByIdNotNull(Pageable pageable);
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/repository/BookRepository.java]: Can ID be null?

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! But fix my comments before merging  ;)

**Pull request merged**